File: AgencyCampaign/AgencyCampaign.Domain/Entities/FinancialEntry.cs
using AgencyCampaign.Domain.ValueObjects;
using Archon.Core.Entities;

namespace AgencyCampaign.Domain.Entities
{
    public sealed class FinancialEntry : Entity
    {
        public long AccountId { get; private set; }

        public FinancialAccount? Account { get; private set; }

        public long? CampaignId { get; private set; }

        public Campaign? Campaign { get; private set; }

        public long? CampaignDeliverableId { get; private set; }

        public CampaignDeliverable? CampaignDeliverable { get; private set; }

        public FinancialEntryType Type { get; private set; }

        public FinancialEntryCategory Category { get; private set; }

        public string Description { get; private set; } = string.Empty;

        public decimal Amount { get; private set; }

        public DateTimeOffset DueAt { get; private set; }

        public DateTimeOffset OccurredAt { get; private set; }

        public string? PaymentMethod { get; private set; }

        public string? ReferenceCode { get; private set; }

        public DateTimeOffset? PaidAt { get; private set; }

        public FinancialEntryStatus Status { get; private set; } = FinancialEntryStatus.Pending;

        public string? CounterpartyName { get; private set; }

        public string? Notes { get; private set; }

        private FinancialEntry()
        {
        }

        public FinancialEntry(long accountId, FinancialEntryType type, FinancialEntryCategory category, string description, decimal amount, DateTimeOffset dueAt, DateTimeOffset occurredAt, string? paymentMethod = null, string? referenceCode = null, string? counterpartyName = null, string? notes = null, long? campaignId = null, long? campaignDeliverableId = null)
        {
            ArgumentOutOfRangeException.ThrowIfNegativeOrZero(accountId);
            ArgumentException.ThrowIfNullOrWhiteSpace(description);
            ArgumentOutOfRangeException.ThrowIfNegative(amount);

            AccountId = accountId;
            CampaignId = campaignId;
            CampaignDeliverableId = campaignDeliverableId;
            Type = type;
            Category = category;
            Description = description.Trim();
            Amount = amount;
            DueAt = dueAt.ToUniversalTime();
            OccurredAt = occurredAt.ToUniversalTime();
            PaymentMethod = Normalize(paymentMethod);
            ReferenceCode = Normalize(referenceCode);
            CounterpartyName = Normalize(counterpartyName);
            Notes = Normalize(notes);
        }

        public void Update(long accountId, FinancialEntryType type, FinancialEntryCategory category, string description, decimal amount, DateTimeOffset dueAt, DateTimeOffset occurredAt, string? paymentMethod, string? referenceCode, string? counterpartyName, string? notes, long? campaignId, long? campaignDeliverableId)
        {
            ArgumentOutOfRangeException.ThrowIfNegativeOrZero(accountId);
            ArgumentException.ThrowIfNullOrWhiteSpace(description);
            ArgumentOutOfRangeException.ThrowIfNegative(amount);

            AccountId = accountId;
            CampaignId = campaignId;
            CampaignDeliverableId = campaignDeliverableId;
            Type = type;
            Category = category;
            Description = description.Trim();
            Amount = amount;
            DueAt = dueAt.ToUniversalTime();
            OccurredAt = occurredAt.ToUniversalTime();
            PaymentMethod = Normalize(paymentMethod);
            ReferenceCode = Normalize(referenceCode);
            CounterpartyName = Normalize(counterpartyName);
            Notes = Normalize(notes);
        }

        public void ChangeStatus(FinancialEntryStatus status, DateTimeOffset? paidAt = null)
        {
            Status = status;
            PaidAt = status == FinancialEntryStatus.Paid ? paidAt?.ToUniversalTime() : null;
        }

        public void RecalculateOverdue(DateTimeOffset now)
        {
            if (Status == FinancialEntryStatus.Pending && DueAt < now)
            {
                Status = FinancialEntryStatus.Overdue;
                return;
            }

            if (Status == FinancialEntryStatus.Overdue && DueAt >= now)
            {
                Status = FinancialEntryStatus.Pending;
            }
        }

        private static string? Normalize(string? value)
        {
            return string.IsNullOrWhiteSpace(value) ? null : value.Trim();
        }
    }
}

File: AgencyCampaign/AgencyCampaign.Web/src/types/financialEntry.ts
export const FinancialEntryType = {
  Receivable: 1,
  Payable: 2,
} as const

export type FinancialEntryTypeValue = (typeof FinancialEntryType)[keyof typeof FinancialEntryType]

export const FinancialEntryStatus = {
  Pending: 1,
  Paid: 2,
  Overdue: 3,
  Cancelled: 4,
} as const

export type FinancialEntryStatusValue = (typeof FinancialEntryStatus)[keyof typeof FinancialEntryStatus]

export const financialEntryStatusLabels: Record<number, string> = {
  1: 'Pendente',
  2: 'Pago',
  3: 'Vencido',
  4: 'Cancelado',
}

export const financialEntryReceivableStatusLabels: Record<number, string> = {
  1: 'A receber',
  2: 'Recebido',
  3: 'Vencido',
  4: 'Cancelado',
}

export const financialEntryCategoryLabels: Record<number, string> = {
  1: 'Recebível da marca',
  2: 'Repasse creator',
  3: 'Fee da agência',
  4: 'Custo operacional',
  5: 'Bônus',
  6: 'Ajuste',
  7: 'Reembolso',
  8: 'Imposto',
}

export interface FinancialEntry {
  id: number
  accountId: number
  accountName?: string | null
  accountColor?: string | null
  campaignId?: number | null
  campaignName?: string | null
  campaignDeliverableId?: number
  type: number
  category: number
  description: string
  amount: number
  dueAt: string
  occurredAt: string
  paymentMethod?: string
  referenceCode?: string
  paidAt?: string
  status: number
  counterpartyName?: string
  notes?: string
  createdAt: string
  updatedAt?: string
}

export interface FinancialSummary {
  type: number
  totalPending: number
  totalSettledThisMonth: number
  totalOverdue: number
  totalDueNext7Days: number
  pendingCount: number
  overdueCount: number
}
